Add "in progress" state to RuntimeInformation

RobinTF opened the pull request:
* Mark the currently running operation as well as all its ancestors in the runtime tree as `in progress` (previously they were shown as `not started`. Note that it is easy to filter out the "actually" running operation by traversing the runtime information.
* Fix a bug in the runtime information that the time of the currently running operation was negative. It is now zero.\
* Also refactor the `RuntimeInformation` and `Operation` class to consistently use types from `std::chrono` for durations.

joka921 reviewed (commented):
The functionality is fine, 
but maybe we can use this PR to add (at least initial) tests for the runtime info updates inside the `OperationTest.cpp` file,
(We have several dummy operations for testing, for example `ValuesForTesting`.

For this PR at least the following two workflows would be nice:

- A succesfully computed result sets a inProgress status and then a done status.
- A result that is read from cache (just retrigger the computation) does never set inProgress.
(I hope this is doable by changing the queryUpdate mechanism to a custom mock that exposes the runtime information to a test).
If you want to, you can also add the tests for the other cases, s.t. we add the unit tests for "we always get the updates when something changes, and they contain the desired contents".

Also:
Do you plan to perform the "filtering" for the "actually active" computation in the UI? I could live with that (it is not functionally important), I just wanted to make sure that that is on purpose not part of this PR here.

RobinTF commented:
> Do you plan to perform the "filtering" for the "actually active" computation in the UI? I could live with that (it is not functionally important), I just wanted to make sure that that is on purpose not part of this PR here.

Yes, this is also the reason why I decided to use the wording "in progress" here. Of course this is a topic up for debate, but in this case I decided to consider an operation that is currently waiting for a child to be computed also as being "in progress" of being computed. The UI can than decide to highlight all of them or just a single one.
This makes keeps the code simpler in my opinion. No need to introduce complexity just for a rendering "issue".

Also speaking of it, I should also include the bugfix to this PR that prevents the totalTime from being negative at times.

I agree on the tests.

RobinTF commented:
@joka921 I added the tests, and added the fix to this PR that prevents negative timing values.
While doing this I (again) stumbled above the double that stores the duration of the computation time, which I found odd. So while I was already at it, I also refactored the code to use std::chrono types instead, which revealed that only the tests were ever making use of fractions, non-test code always used the double like any other int + rounding errors and verbose static casting.

RobinTF commented on test/RuntimeInformationTest.cpp:
```diff
@@ -7,52 +7,55 @@
   parent.children_.push_back(std::make_shared<RuntimeInformation>(child2));
 
-  // 2.5 == 10.0 - 4.5 - 3.0
-  ASSERT_DOUBLE_EQ(parent.getOperationTime(), 2.5);
```
I don't think this test was working as intended for a while

joka921 commented on src/engine/Operation.cpp:
```diff
@@ -284,7 +287,7 @@ void Operation::updateRuntimeInformationWhenOptimizedOut(
                              const auto& self) -> void {
     rti.status_ = status;
-    rti.totalTime_ = 0;
+    rti.totalTime_ = millis::zero();
```
```suggestion
    rti.totalTime_ = 0ms;
```
(you need a using namespace std::chrono_literals or whatever it is called)
(also use consistently at the other places).

joka921 commented on src/engine/RuntimeInformation.cpp:
```diff
@@ -66,18 +66,20 @@ void RuntimeInformation::writeToStream(std::ostream& out, size_t indent) const {
-        << " ms" << '\n';
     out << indentStr(indent)
-        << "original_operation_time: " << originalOperationTime_ << " ms"
+        << "original_total_time: " << originalTotalTime_.count() << " ms"
```
Add a comment "`TODO<g++12, Clang 17>` use `<< originalTotalTime_` directly".
(It is part of C++20, but not all of our current c ompilers support it.

joka921 commented on src/engine/RuntimeInformation.cpp:
```diff
@@ -122,8 +124,11 @@ double RuntimeInformation::getOperationTime() const {
+    // Prevent "negative" computation times in case totalTime_ was not
+    // computed for this yet.
+    constexpr auto zero = std::chrono::milliseconds::zero();
+    return std::max(zero, -std::reduce(timesOfChildren.begin(),
```
Use `0ms` literal.
And the `std::reduce` with the double negative is really unreadable here, please use the direct calculation as before....

joka921 commented on src/engine/RuntimeInformation.h:
```diff
@@ -39,12 +39,14 @@ class RuntimeInformation {
   /// The total time spent computing this operation. This includes the
   /// computation of the children.
-  double totalTime_ = 0.0;
+  std::chrono::milliseconds totalTime_ = std::chrono::milliseconds::zero();
```
0ms

RobinTF replied:
Unfortunately we can't use `using namespace` within a class without cluttering the global namespace. I went with an alternative that is slightly better than this

joka921 commented on src/engine/RuntimeInformation.h:
```diff
@@ -39,12 +39,14 @@ class RuntimeInformation {
+  std::chrono::milliseconds originalTotalTime_ =
+      std::chrono::milliseconds::zero();
+  std::chrono::milliseconds originalOperationTime_ =
+      std::chrono::milliseconds::zero();
```
same.

joka921 commented on src/engine/RuntimeInformation.h:
```diff
@@ -91,10 +93,10 @@ class RuntimeInformation {
   /// Get the time spent computing the operation. This is the total time minus
   /// the time spent computing the children.
-  [[nodiscard]] double getOperationTime() const;
+  [[nodiscard]] std::chrono::milliseconds getOperationTime() const;
```
This is the function you changed, so you changed the wrong docstring...

RobinTF replied:
Oh right, good catch

joka921 commented on src/engine/RuntimeInformation.h:
```diff
@@ -109,7 +111,8 @@ class RuntimeInformation {
   // runtime information) is written to the cache, are passed in as arguments.
-  void addLimitOffsetRow(const LimitOffsetClause& l, size_t timeForLimit,
+  void addLimitOffsetRow(const LimitOffsetClause& l,
+                         std::chrono::milliseconds timeForLimit,
```
Maybe for readability you can also introduce `RuntimeInformation::Milliseconds` for shorter lines.

joka921 commented on src/index/CompressedRelation.cpp:
```diff
@@ -146,10 +146,14 @@ CompressedRelationReader::asyncParallelBlockGenerator(
+  auto storeTime = [&details, &popTimer]() {
+    details.blockingTime_ =
+        std::chrono::duration_cast<std::chrono::milliseconds>(popTimer.value());
+  };
```
This also gets much simpler again once we can directly get the milliseconds from the timer (it is the "typical" resolution for all of our use cases anywayl.

joka921 commented on src/index/CompressedRelation.cpp:
```diff
@@ -164,7 +168,7 @@ CompressedRelationReader::asyncParallelBlockGenerator(
   // The `OnDestruction...` above might be called too late, so we manually set
   // the timer again.
-  details.blockingTimeMs_ = popTimer.msecs();
+  storeTime();
```
I think we either have to stop the timer before this call (then we can get rid of the whole block), or we have to cancel the Cleanup. Otherwise we probably overwrite the runtime info again with a possibly "too large" value.
This is a bug of mine.
Maybe just replace this line by `popTimer.stop()` and get rid of the comment and the explicit storeTime.

RobinTF replied:
Right, didn't consider the case where the timer wasn't stopped. In all other cases it would just redundantly store the same value twice.

joka921 commented on src/index/CompressedRelation.h:
```diff
@@ -259,7 +259,7 @@ class CompressedRelationReader {
     size_t numBlocksAll_ = 0;
     size_t numElementsRead_ = 0;
-    size_t blockingTimeMs_ = 0;
+    std::chrono::milliseconds blockingTime_ = std::chrono::milliseconds::zero();
```
0 ms.

RobinTF replied:
Can't do without adding it to the global namespace

joka921 replied:
This problem probably will also be used by C++20 modules I guess
(Cluttering the global namespace not only with macros, but also with these very useful literal namespaces

joka921 commented on test/RuntimeInformationTest.cpp:
```diff
@@ -7,52 +7,55 @@
   RuntimeInformation rti;
   rti.descriptor_ = "BaseOperation";
-  rti.totalTime_ = 4.0;
+  rti.totalTime_ = milliseconds{4};
```
`4ms` (chrono literals, very useful for exactly this case).

joka921 commented on test/OperationTest.cpp:
```diff
@@ -79,3 +83,67 @@ TEST(OperationTest, getResultOnlyCached) {
+
+// _____________________________________________________________________________
+
+MATCHER_P2(HasJsonKeyValue, key, value, "") {
+  try {
+    auto json = nlohmann::json::parse(arg);
+    if (!json.contains(key)) {
+      *result_listener << "JSON object does not contain key \"" << key << '"';
+      return false;
+    }
+    if (json[key] == value) {
+      return true;
+    }
+    *result_listener << "JSON key \"" << key << "\" has value \"" << json[key]
+                     << '"';
+  } catch (const nlohmann::json::parse_error& error) {
+    *result_listener << "failed to parse JSON";
+  }
+  return false;
+}
```
This matcher is too generic and probably useful to let it go to waste in the `.cpp` file
Maybe add it to the `GTestHelpers.h` file (where we also have the generic `AD_EXPECT_THROW_WITH_MESSAGE and add a short comment and example.

joka921 commented on test/OperationTest.cpp:
```diff
@@ -79,3 +83,67 @@ TEST(OperationTest, getResultOnlyCached) {
+  EXPECT_THAT(jsonHistory,
+              ElementsAre(HasJsonKeyValue("status", "not started"),
+                          HasJsonKeyValue("status", "in progress"),
+                          HasJsonKeyValue("status", "fully materialized"),
```
comment why the `fully materialized` appears twice, that is not clear at first glance. (And it is also something that is not a bug if changed in the future, so state here that this is just an artifact of the current implementation.

joka921 commented on test/OperationTest.cpp:
```diff
@@ -79,3 +83,67 @@ TEST(OperationTest, getResultOnlyCached) {
+
+  EXPECT_THAT(jsonHistory,
+              ElementsAre(HasJsonKeyValue("status", "fully materialized"),
+                          HasJsonKeyValue("status", "fully materialized")));
```
A lot of common boilerplate for both tests. 
Best add the common code (setting up of index and qec + history object + operation to a separate helper function or fixture or whatever, s.t. its easier to add the remaining tests in the similar manner that are also missing, but not related to this PR.

joka921 reviewed (changes requested):
Very nice, this is now much clearer, I have several suggestions all of which are just minor code quality tweaks.

RobinTF commented on src/engine/Operation.cpp:
```diff
@@ -8,6 +8,10 @@
 #include "util/OnDestructionDontThrowDuringStackUnwinding.h"
 #include "util/TransparentFunctors.h"
 
+std::chrono::milliseconds toMs(const ad_utility::Timer& timer) {
```
This is something for a separate PR. CLion reports 24 other usages all across the codebase

joka921 commented:
Thanks for checking, can you setup that pr after this one is finished?

RobinTF ***@***.***> schrieb am Fr., 3. Nov. 2023, 18:19:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/engine/Operation.cpp
> <https://github.com/ad-freiburg/qlever/pull/1131#discussion_r1382002658>:
>
> > @@ -8,6 +8,10 @@
>  #include "util/OnDestructionDontThrowDuringStackUnwinding.h"
>  #include "util/TransparentFunctors.h"
>
> +std::chrono::milliseconds toMs(const ad_utility::Timer& timer) {
>
> This is something for a separate PR. CLion reports 24 other usages all
> across the codebase
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ad-freiburg/qlever/pull/1131#discussion_r1382002658>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADS6SAXH5PT74RH5EPMCQWLYCURRNAVCNFSM6AAAAAA62AYM5OVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOMJTGE4TEOJZGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

RobinTF commented:
@joka921 
> Thanks for checking, can you setup that pr after this one is finished?

Yes, but it will probably land after #1125 because this PR touches the timer class too (trying to minimize merge conflicts). Also I plan to remove the usage of `std::chrono::high_resolution_clock` in favour of `std::chrono::steady_clock`, because that depending on the standard lib this has different behaviour with regard to clock changes of the system.
